### PR TITLE
Removed usage of deprecated ContainerAwareEventDispatcher

### LINKED ISSUE #247: ContainerAwareEventDispatcher deprecated
## Subject

Starting Symfony the use of Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher is deprecated and will give the following warning: 
The Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher class is deprecated since version 3.3 and will be removed in 4.0. Use EventDispatcher with closure factories instead:

It would be beneficial if this old interface could be swapped with the new one

## Steps to reproduce

Install sonata notification with Symfony 3.3

## Expected results

No warnings on unit tests

## Actual results

The Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher class is deprecated since version 3.3 and will be removed in 4.0. Use EventDispatcher with closure factories instead:

### PR DESCRIPTION
**Pull request by KDederichs**
I am targeting this branch, because I removed the deprecated ContainerAwareEventDispatcher and according to your contributing guide that goes here .

Closes #247

## Changelog

```markdown
### Fixed
- Use `EventDispatcher` instead of deprecated `ContainerAwareEventDispatcher`
```

## Subject
As of Symfony 3.3 the ContainerAwareEventDispatcher is deprecated and throws deprecation notices during Unit tests.
This PR fixes the deprecation by using the proposed EventDispatcher

### CODE REVIEW

**Comment by greg0ire** [comment]
Can you make a BC PR that uses the new way as soon as it is available, and the old way otherwise?

**Comment by KDederichs** [comment]
Not sure what you mean, according to the deprecation notice the EventDispatcher is the new way to go which I did use here.

**Comment by greg0ire** [comment]
I mean that if we detect our user is using sf 3.2, we should not use the new way but the old one. Make it conditional.

**Comment by KDederichs** [comment]
Ok lets see what I can come up with

File: DependencyInjection/Compiler/NotificationCompilerPass.php [hunk 1]
```diff
@@ -50,14 +51,25 @@ public function process(ContainerBuilder $container)
...
 
-                $definition->addMethodCall(
-                    'addListener',
-                    array(
-                        $event['type'],
-                        array(new ServiceClosureArgument(new Reference($id)), 'process'),
-                        $priority,
-                    )
-                );
+                if (Kernel::MAJOR_VERSION <= 3 && Kernel::MINOR_VERSION < 3) {
```
Comment (OskarStark) [inline DependencyInjection/Compiler/NotificationCompilerPass.php]: couldn't we check here if `addListenerService` is callable

File: Factory/EventDispatcherFactory.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+namespace Sonata\NotificationBundle\Factory;
+
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\HttpKernel\Kernel;
+
+class EventDispatcherFactory
+{
+    public static function createEventDispatcher(ContainerInterface $serviceContainer)
```
Comment (OskarStark) [inline Factory/EventDispatcherFactory.php]: @greg0ire `$container` or `$serviceContainer` naming?

File: DependencyInjection/Compiler/NotificationCompilerPass.php (hunk 1 above)
Comment (greg0ire) [inline DependencyInjection/Compiler/NotificationCompilerPass.php]: Please read this OP : https://symfony.com/doc/current/setup/bundles.html#updating-your-code-to-support-symfony-2-x-and-3-x-at-the-same-time

File: Factory/EventDispatcherFactory.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+namespace Sonata\NotificationBundle\Factory;
+
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\HttpKernel\Kernel;
+
+class EventDispatcherFactory
```
Comment (OskarStark) [inline Factory/EventDispatcherFactory.php]: `@author` is missing

**Comment by greg0ire** [comment]
I this is going to be BC I am rebasing this on 3.x

**Comment by greg0ire** [comment]
1. Fetch all new commits: `git fetch --all`. ⬇️ 
2. Rebase on what you fetched interactively: ✂️ 
    1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin repository.
    2. A window will show up with many lines, remove every line but the last one, which should correspond to your commit.
   2. If you run into a conflict: 💥 
        1. fix it with `git mergetool`. 😷 
        1. continue on your merry way with `git rebase --continue`. ⏩ 
3. Force push to overwrite all this : `git push --force`. ⬆️

File: DependencyInjection/Compiler/NotificationCompilerPass.php (hunk 1 above)
Comment (KDederichs) [inline DependencyInjection/Compiler/NotificationCompilerPass.php]: Ok then let me do that, once sec

File: Factory/EventDispatcherFactory.php (hunk 3 above)
Comment (greg0ire) [inline Factory/EventDispatcherFactory.php]: Also, this should be final

Comment (greg0ire) [inline Factory/EventDispatcherFactory.php]: Hmmm, and I'm not a huge fan of factories... how about using dependency injection instead? I mean, the `event_dispatcher` is already available as a service, maybe the other one is too? So you would just have to make the service definition dynamic.

File: Factory/EventDispatcherFactory.php (hunk 2 above)
Comment (greg0ire) [inline Factory/EventDispatcherFactory.php]: `$container`, but I have doubts about the future of this class.

File: Factory/EventDispatcherFactory.php [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\HttpKernel\Kernel;
+
+final class EventDispatcherFactory
+{
+    public static function createEventDispatcher(ContainerInterface $container)
+    {
+        if (Kernel::MAJOR_VERSION <= 3 && Kernel::MINOR_VERSION < 3) {
```
Comment (core23) [inline Factory/EventDispatcherFactory.php]: Don't rely on the kernel version. We have no dependency on the full symfony stack.

File: Factory/EventDispatcherFactory.php (hunk 3 above)
Comment (greg0ire) [inline Factory/EventDispatcherFactory.php]: ping @KDederichs can you give this a try?

Comment (KDederichs) [inline Factory/EventDispatcherFactory.php]: It seems to me like event_dispatcher is the only service declaration and even 3.3 that still is declared using ContainerAwareEventDispatcher in the Symofny FrameworkBundle

Comment (greg0ire) [inline Factory/EventDispatcherFactory.php]: Not sure I'm following... are you telling me they still use the deprecated way themselves? that's very unlikely. Or maybe the service is deprecated in favor of another...

**Comment by greg0ire** [comment]
The deprecating PR, for reference: https://github.com/symfony/symfony/pull/20937

File: Factory/EventDispatcherFactory.php (hunk 3 above)
Comment (KDederichs) [inline Factory/EventDispatcherFactory.php]: Looks like that to me:
https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml <-- Services in 3.2

https://github.com/symfony/symfony/blob/3.3/src/Symfony/Bundle/FrameworkBundle/Resources/config/services.xml <-- Services in 3.3

Both have event_dispatcher  as ContainerAwareEventDispatcher

**Comment by greg0ire** [comment]
After reading https://github.com/symfony/symfony/issues/22402, I understand your PR better.
The issue I have with that PR is that dropping compatibility will be a bit complicated.
Can you change `core.xml` so that it corresponds to the new way, and create a `core_legacy.xml` file, and choose which file to load with the condition you found? That way when dropping compat, it will be very simple:

- delete `core_legacy.xml`
- remove condition

File: DependencyInjection/Compiler/NotificationCompilerPass.php [hunk 5]
```diff
@@ -48,14 +50,25 @@ public function process(ContainerBuilder $container)
...
 
-                $definition->addMethodCall(
-                    'addListenerService',
-                    array(
-                        $event['type'],
-                        array($id, 'process'),
-                        $priority,
-                    )
-                );
+                if (!class_exists('Symfony\Component\DependencyInjection\Argument\ServiceClosureArgument')) {
```
Comment (greg0ire) [inline DependencyInjection/Compiler/NotificationCompilerPass.php]: Please add `NEXT_MAJOR` instructions here

File: Resources/config/core_legacy.xml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (greg0ire) [inline Resources/config/core_legacy.xml]: And here too, of course

**Comment by greg0ire** [comment]
Please fix StyleCI

**Comment by KDederichs** [comment]
Fixed it, and everything seems to have passed

**Comment by greg0ire** [comment]
@core23 @OskarStark please review again

**Comment by OskarStark** [comment]
Please check for existance and not like this:

```php
if (!class_exists('Symfony\Component\DependencyInjection\Argument\ServiceClosureArgument')) {
```

like you do in the compiler pass.

**Comment by KDederichs** [comment]
done @OskarStark

**Comment by greg0ire** [comment]
Thanks @KDederichs !

**Pull request merged**